File: collge_major/2nd-year/semester-1/data-structures-and-algorithms/data-structures-and-algorithm-design/data-structures/queue/queue.js
class Queue {
  constructor() {
    this.queue = [];
  }

  enqueue(element) {
    // add element
    return this.queue.push(element);
  }

  dequeue() {
    if (this.queue.length > 0) {
      return this.queue.shift(); // remove first element
    }
  }

  peek() {
    return this.queue[0];
  }

  size() {
    return this.queue.length;
  }

  isEmpty() {
    return this.queue.length == 0;
  }

  clear() {
    this.queue = [];
  }
}

let counter = new Queue();

// add
counter.enqueue(1);
counter.enqueue(2);
counter.enqueue(3);
counter.enqueue(4);
counter.enqueue(5);

// queue
console.log(`Queue: [${counter.queue}]`);

// get last element
console.log(`Peek: ${counter.peek()}`); // 3

// remove
counter.dequeue(); // 1
counter.dequeue(); // 2

console.log(`After Dequeuing the 1st and 2nd element`);
console.log(`Queue: [${counter.queue}]`);

console.log(`is Empty: ${counter.isEmpty()}`); // false
console.log(`Size: ${counter.size()}`); // 1

// counter.clear();
// console.log(counter.size()); // 0

/*
 - Complexity of queue operations
 - We are always going to access the first queue element and add new elements at the end of the queue. To do this, we will use:

 - for enqueue – O(1)
 - for dequeue – O(1)


********/

File: collge_major/2nd-year/semester-1/data-structures-and-algorithms/data-structures-and-algorithm-design/data-structures/queue/queue.cpp
#include <iostream>
using namespace std;

class Queue
{
    int front, rear, capacity;
    int *queue;

public:
    Queue(int c)
    {
        front = rear = 0;
        capacity = c;
        queue = new int;
    }
    ~Queue() { delete[] queue; }

    void enqueue(int data)
    {
        if (capacity == rear)
        {
            cout << "Queue is full\n";
            return;
        }
        else
        {
            queue[rear] = data;
            rear++;
        }
    }
    void dequeue()
    {
        if (front == rear)
        {
            cout << "Queue is empty\n";
            return;
        }
        else
        {
            for (int i = 0; i < rear - 1; i++)
            {
                queue[i] = queue[i + 1];
            }
            rear--;
        }
    }

    int peek()
    {
        if (front == rear)
        {
            cout << "Queue is empty\n";
            return -1;
        }
        return queue[front];
    }

    bool isEmpty()
    {
        return front == rear;
    }
};

int main()
{
    Queue q(5);

    // Demonstrate enqueue operation
    q.enqueue(10);
    q.enqueue(20);
    q.enqueue(30);

    // Demonstrate dequeue operation
    q.dequeue();

    // Check front element
    cout << "Front element is: " << q.peek() << endl;

    // Check if queue is empty
    cout << "Queue empty: " << (q.isEmpty() ? "True" : "False") << endl;

    return 0;
}